design/design-doc-external-secret-stores.md

```diff
@@ -0,0 +1,416 @@
+configuration. This would register the secret store during initialization and
+will be available to publish connection details if requested via resource APIs.
+
+**Plugins for different secret stores**
+
+These plugins would implement a well known API and serve with gRPC.
+Communication and authentication with the Secret Store would be the
+responsibility of this component. Plugins would authenticate to external secret
+stores using a configuration file or environment variables.
+
+**Mutating webhook injecting plugin sidecars with proper configuration**
+
+When enabled/configured, this component would be responsible for injecting
+plugin sidecars with proper configuration.
+
+##### API
+
+API will be similar to option 1 but not the same since we would no longer be
+able to have strong types per secret store option. In the API, users would
+specify the name of the secret store which should match the registered store
+name and arbitrary key value pairs which would be required by the selected
+secret store. Store `kubernetes` would be available by default which would
+provide previous behaviour to keep connection details in Kubernetes secrets.
+
+```
+apiVersion: widgets.crossplane.io/v1beta1
+kind: Example
+metadata:
+  name: my-cool-managed-resource
+spec:
+  forProvider:
+    coolness: 24
+  publishConnectionDetailsTo:
+  - store: kubernetes
+      secretName: my-cool-managed-resource-connection
+      secretNamespace: secrets
+  - store: vault
+      path: /path/to/secret/my-cool-managed-resource-connection
+```
+
+##### Example Deployment for Using Vault as Secret Store
+
+1. User deploys [Vault sidecar injector] to the cluster.
+2. User configures proper Vault authentication, e.g. Kubernetes auth
+3. User deploys **Crossplane Secret Store Plugin Webhook** with proper
+configuration:
+   1. Annotate with:
+      ```
+      vault.hashicorp.com/agent-inject: "true"
+      vault.hashicorp.com/agent-inject-token: "true"
+      ```
+   2. Enable Vault plugin with:
+      ```
+      VAULT_ADDRESS=https://vault-cluster.vault-system:8200
+      VAULT_TOKEN_FILE=/vault/secrets/token
+      ```
+4. User deploys Crossplane and provider controllers.
+5. **Crossplane Secret Store Plugin Webhook** injects plugin sidecar and also
+registers plugin by passing plugin config flags to crossplane and provider
+controllers.
+6. Crossplane and provider controllers registers Vault as a secret store at
+startup.
+7. Crossplane and provider controllers would publish connection details to Vault
+if requested in the managed resource API (`publishConnectionDetailsTo` includes
+`Vault` as store).
+
+![Secret Store Plugin Example][secret-store-plugin-example]
+
+#### Option 3: Propagate from K8S secrets
+
+_In this option we assume that it is ok to keep storing connection details as
+Kubernetes secrets (temporarily or permanently) as long as they will also be
+propagated to external secret stores if configured/requested._
```

Comment:
Comment from @negz 

This is my preferred approach from an architectural perspective, but I think we will find that we can't make this assumption. I'm pretty sure this option will be a non-starter because anecdotally I've heard multiple orgs state that their security teams deem Kubernetes Secrets to be insecure, and thus they can't be used to store sensitive data at all.

Reply:
@muvaf I came here to bring this up and see that you beat me to it :) I have a working demo of this, and you are right that at first glance the read operation is not as clean as the write. That being said, I worked around this by changing composition to not actually read the managed resource `Secret`, but rather write the "composed" `Secret` with annotations that inform the webhook how to aggregate a new external secret from the managed resource external secrets it already knows about. I believe this was working well, and it actually reduces some of our permissions surface area because core Crossplane doesn't ever need to actually know the contents of any managed resource `Secret` it just instructs the mutating webhook to use another the contents of another `Secret` it knows about compose a new one.

Reply:
@hasheddan But we have a restriction that the sensitive information shouldn't end up in a secret. If I understood your suggestion correctly, there will be a secret with content, right?

Reply:
@muvaf there will not be -- the webhook just has a notion of the mapping between the (empty) managed resource `Secrets` and their external representations. Composition machinery creates a composite `Secret` that has annotations referencing the (empty) managed resource `Secrets`, and the webhook intercepts and composes the new external composed secret by fetching the external managed resource secret data and combining them appropriately. The caveat here is that composition would have to be okay with writing an empty composite `Secret` when the managed resource `Secrets` are also empty.

Reply:
@hasheddan that's a neat approach. 
I am definitely happy to discuss further and am wondering if we could make this completely transparent to Crossplane, requiring no changes in composition logic at all. I believe this is not possible because webhooks cannot intercept read requests but wondering if we could extend that approach (intercepting secret requests) with a k8s API server proxy (similar to how we did in upbound agent proxy) instead of webhooks 🤔 
This would be a solution that could work for any Kubernetes application and not only for Crossplane.

Reply:
One thing I'm curious about with this approach is how folks would find the UX. Wouldn't (e.g) 'writeConnectionSecretToRef' be a little confusing if we're actually never writing a Secret? My hunch is that our types declaring the _actual_ secret sink in their APIs would be more intuitive and self documenting.

This problem arguably also exists in the case in which we're writing a Secret that is then propagated to some other system, but somehow the fact that the 'intermediary' Secret actually exists makes that seem more tolerable to me.

Reply:
That all said - this Secret intercept idea does sound cool and may be worth proving out further even if we don't use it in this specific context.

Reply:
> I believe this is not possible because webhooks cannot intercept read requests but wondering if we could extend that approach (intercepting secret requests) with a k8s API server proxy (similar to how we did in upbound agent proxy) instead of webhooks

@turkenh this is an interesting approach that I had also considered, but I think we can alleviate the need to do so with writing an empty `Secret` with annotations as mentioned, which would mean that Crossplane (composition) would never try to read a `Secret`, it would just be essentially creating "secret pointers" that the webhook knew how to resolve.

> Wouldn't (e.g) 'writeConnectionSecretToRef' be a little confusing if we're actually never writing a Secret? (Or rather writing an empty one.) 

@negz this could absolutely be the case, and IMO is the biggest issue with this direction. However, I do think it is possible that it could actually be useful to have a marker that says "this thing exists externally" as it kind of creates a similar abstraction layer over secret stores that managed resources do over external APIs. We could make this even more clear by having the mutating webhook add an annotation that indicates what secret store backend the secret contents were actually written to, as well as (potentially) what fields are present in external secret.